lib/workers/pr/changelog/source-gitlab.ts

```diff
@@ -0,0 +1,163 @@
+import * as globalCache from '../../../util/cache/global';
+import { regEx } from '../../../util/regex';
+import * as allVersioning from '../../../versioning';
+import { BranchUpgradeConfig } from '../../common';
+import { ChangeLogRelease, ChangeLogResult } from './common';
+import { addReleaseNotes } from './release-notes';
+
+const { get: glGot } = api;
+
+async function getTags(
+  endpoint: string,
+  versionScheme: string,
+  repository: string
+): Promise<string[]> {
+  logger.trace('getTags() from gitlab');
+  let url = endpoint
+    ? endpoint.replace(/\/?$/, '/')
+    : /* istanbul ignore next: not possible to test, maybe never possible? */ 'https://gitlab.com/api/v4/';
+  const repoid = repository.replace(/\.git/, '').replace(/\//g, '%2F');
+  url += `projects/${repoid}/repository/tags`;
+  try {
+    const res = await glGot<{ name: string }[]>(url, {
+      paginate: true,
+    });
+
+    const tags = (res && res.body) || [];
+
+    if (!tags.length) {
+      logger.debug({ repoid }, 'repository has no Gitlab tags');
+    }
+
+    return tags.map((tag) => tag.name).filter(Boolean);
+  } catch (err) {
+    logger.info({ sourceRepo: repoid }, 'Failed to fetch Gitlab tags');
+    // istanbul ignore if
+    if (err.message && err.message.includes('Bad credentials')) {
+      logger.warn('Bad credentials triggering tag fail lookup in changelog');
+      throw err;
+    }
+    return [];
+  }
+}
+
+export async function getChangeLogJSON({
+  endpoint,
+  versioning,
+  fromVersion,
+  toVersion,
+  sourceUrl,
+  releases,
+  depName,
+  manager,
+}: BranchUpgradeConfig): Promise<ChangeLogResult | null> {
+  if (sourceUrl === 'https://gitlab.com/DefinitelyTyped/DefinitelyTyped') {
+    logger.debug('No release notes for @types');
+    return null;
+  }
+  logger.trace('getChangeLogJSON for gitlab');
+  const version = allVersioning.get(versioning);
+  const { protocol, host, pathname } = URL.parse(sourceUrl);
+  logger.debug({ protocol, host, pathname }, 'Protocol, host, pathname');
+  const baseUrl = 'https://gitlab.com/';
+  const apiBaseUrl = 'https://gitlab.com/api/v4/';
+  const repository = pathname.slice(1).replace(/\/$/, '').replace(/\.git/, '');
+  if (repository.split('/').length !== 2) {
+    logger.info({ sourceUrl }, 'Invalid gitlab (github) URL found');
+    return null;
+  }
+  if (!(releases && releases.length)) {
+    logger.debug('No releases');
+    return null;
+  }
+  // This extra filter/sort should not be necessary, but better safe than sorry
+  const validReleases = [...releases]
+    .filter((release) => version.isVersion(release.version))
+    .sort((a, b) => version.sortVersions(a.version, b.version));
+
+  if (validReleases.length < 2) {
+    logger.debug('Not enough valid releases');
+    return null;
+  }
+
+  let tags: string[];
+
+  async function getRef(release: Release): Promise<string | null> {
+    if (!tags) {
+      tags = await getTags(endpoint, versioning, repository);
```

Comment:
Shouldn't `endpoint` be constructed from sourceUrl?

Reply:
As far as I understand - no. There is api request should be used.

Reply:
@rarkins Do you have a suggestion here. It looks like the `getTags` will call the configured global app/cli `endpoint`, but should call the the api where `sourceUrl` points to. 🤔

Reply:
You're right. I've updated this code and now sourceUrl in use.